File: 0101-0200/0129-Sum Root to Leaf Numbers/0129-Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int sum = 0;
        dfs(root, 0, sum);
        return sum;
    }
    
private:
    void dfs(TreeNode* root, int curr, int& sum) {
        if (root == nullptr) {
            return;
        }
        
        curr = curr * 10 + root->val;
        if (root->left == nullptr && root->right == nullptr) {
            sum += curr;
        }
        else {
            dfs(root->left, curr, sum);
            dfs(root->right, curr, sum);
        }
    }
};

File: 0101-0200/0129-Sum Root to Leaf Numbers/0129-Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        def dfs(root, prev):
            if root is None:
                return 0
            curr = prev*10 + root.val
            if root.left is None and root.right is None:
                return curr
            else:
                return dfs(root.left, curr) + dfs(root.right, curr)
        return dfs(root, 0)
